### DISCUSSION TITLE
Dynamic schema component for queries

### DISCUSSION THREAD

**Discussion by cylewitruk** [diesel-rs/diesel#3865] [Q&A]
I'm currently in a case where I want to be able to create the same database schema/structure within separate db schemas (i.e. normal rdbms `CREATE SCHEMA` or SQLite's `ATTACH DATABASE`), which are then usually accessed via `my_schema.my_table` or similar syntax. Ideally I'd like to be able to control the schema being used per `Connection` which is decided at runtime. Had a quick look at the dynamic schema crate but that seems more geared towards dynamic schema/structure.

Any ideas on how I might solve this using the tools currently available?

Thanks!
/ Cyle

**Comment by weiznich** [comment]
Diesel does currently not really provide functionality for switching between schemas at run time. You should be able to use the database URL to specify a specific schema while starting the application. Additionally it's possible to use something like `SET SCHEMA 'your_schema'` at runtime via `sql_query` to manually change the "default" schema while the application is running, although I would expect that this might break some internal caching around custom types.

That written I'm happy to discuss specific additions to the API to make such use case easier, as long as it doesn't break existing use cases.

**Reply by cylewitruk** [reply]
Okay, thanks, then I know!

I think an API addition would probably need to allow setting the schema name at runtime either on the connection or query (with i.e. a `.schema("my_schema")`, since not all databases that Diesel supports support the `set schema 'my_schema'` paradigm. Then each database implementation would need to include that schema name (if present) as a prefix to all objects using that db's DSL for specifying schema name. This because SQLite for example only supports accessing attached databases at query execution (you can't globally set your current schema for attached db's, but each attached db is in a separate file/url).

I would guess a similar method would need to be provided for migration execution.

As you might have guessed by now, my use-case is SQLite :)

**Reply by weiznich** [reply]
I think you might be able to extend the existing query DSL to provide an wrapper that is constructed by calling that `.schema("…")` method on tables. You would need to implement [`QuerySource`](https://docs.diesel.rs/2.1.x/diesel/prelude/trait.QuerySource.html) for some wrapper struct + add quite a few additional trait implementations (`QueryFragment` for constructing the SQL, `QueryId`, `*Dsl` for allowing to add various query dsl releated constructs afterwards, …). That would likely be quite similar to the implementation of `alias!`/`Alias` in diesel: https://docs.diesel.rs/2.1.x/diesel/query_source/struct.Alias.html, so that can be used as an inspiration.

**Reply by cylewitruk** [reply]
Cool, thanks for the tips, I'll have a look when time allows - I've been meaning to spend some time and try to better understand the Diesel internals a little better anyway :smile:
